File: swift-bridge/Sources/WeatherKitBridge/HourlyForecast.swift
import Foundation
import WeatherKit

struct WKHourForecastPayload: Codable {
    let date: String
    let temperature: Double
    let feelsLike: Double
    let condition: String
    let symbolName: String
    let precipitation: String
    let precipitationChance: Double
    let precipitationAmount: Double
    let cloudCover: Double
    let dewPoint: Double
    let humidity: Double
    let isDaylight: Bool
    let pressure: Double
    let pressureTrend: String
    let uvIndex: WKUVIndexPayload
    let visibility: Double
    let wind: WKWindPayload
    let snowfallAmount: Double?
    let cloudCoverByAltitude: WKCloudCoverByAltitudePayload?
}

struct WKHourlyForecastPayload: Codable {
    let forecast: [WKHourForecastPayload]
    let metadata: WKWeatherMetadataPayload
}

func wkPayload(from hourWeather: HourWeather) -> WKHourForecastPayload {
    let snowfallAmount: Double?
    let cloudCoverByAltitude: WKCloudCoverByAltitudePayload?
    if #available(macOS 15.0, *) {
        snowfallAmount = wkLength(hourWeather.snowfallAmount)
        cloudCoverByAltitude = wkPayload(from: hourWeather.cloudCoverByAltitude)
    } else {
        snowfallAmount = nil
        cloudCoverByAltitude = nil
    }

    return WKHourForecastPayload(
        date: wkISO8601String(hourWeather.date),
        temperature: wkTemperature(hourWeather.temperature),
        feelsLike: wkTemperature(hourWeather.apparentTemperature),
        condition: hourWeather.condition.rawValue,
        symbolName: hourWeather.symbolName,
        precipitation: hourWeather.precipitation.rawValue,
        precipitationChance: hourWeather.precipitationChance,
        precipitationAmount: wkLength(hourWeather.precipitationAmount),
        cloudCover: hourWeather.cloudCover,
        dewPoint: wkTemperature(hourWeather.dewPoint),
        humidity: hourWeather.humidity,
        isDaylight: hourWeather.isDaylight,
        pressure: wkPressureValue(hourWeather.pressure),
        pressureTrend: hourWeather.pressureTrend.rawValue,
        uvIndex: wkPayload(from: hourWeather.uvIndex),
        visibility: wkLength(hourWeather.visibility),
        wind: wkPayload(from: hourWeather.wind),
        snowfallAmount: snowfallAmount,
        cloudCoverByAltitude: cloudCoverByAltitude
    )
}

func wkPayload(from hourlyForecast: Forecast<HourWeather>) -> WKHourlyForecastPayload {
    WKHourlyForecastPayload(
        forecast: hourlyForecast.forecast.map { wkPayload(from: $0) },
        metadata: wkPayload(from: hourlyForecast.metadata)
    )
}

@_cdecl("wk_hourly_forecast_retain")
public func wk_hourly_forecast_retain(_ handle: UnsafeMutableRawPointer?) -> UnsafeMutableRawPointer? {
    wkRetainOpaque(handle)
}

@_cdecl("wk_hourly_forecast_release")
public func wk_hourly_forecast_release(_ handle: UnsafeMutableRawPointer?) {
    wkRelease(handle)
}

@_cdecl("wk_hourly_forecast_copy_json")
public func wk_hourly_forecast_copy_json(
    _ handle: UnsafeMutableRawPointer?,
    _ outJSON: UnsafeMutablePointer<UnsafeMutablePointer<CChar>?>?,
    _ outError: UnsafeMutablePointer<UnsafeMutablePointer<CChar>?>?
) -> Int32 {
    wkCopyBoxJSON(handle, outJSON, outError, context: "HourlyForecast", payload: { (value: Forecast<HourWeather>) in wkPayload(from: value) })
}

File: src/hourly_forecast.rs
//! WeatherKit hourly forecast types.

use core::ffi::c_void;

use serde::Deserialize;

use crate::current_weather::{CloudCoverByAltitude, UVIndex, Wind};
use crate::error::WeatherKitError;
use crate::ffi;
use crate::pressure::{deserialize_pressure_trend, Pressure, PressureTrend};
use crate::private::parse_json_from_handle;
use crate::service::WeatherMetadata;
use crate::weather_condition::{
    deserialize_precipitation, deserialize_weather_condition, Precipitation, WeatherCondition,
};

/// Represents a WeatherKit hour forecast.
#[derive(Debug, Clone, PartialEq, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct HourForecast {
    /// Matches the WeatherKit date value.
    pub date: String,
    /// Air temperature in degrees Celsius.
    pub temperature: f64,
    /// Apparent (feels-like) temperature in degrees Celsius.
    pub feels_like: f64,
    /// Matches the WeatherKit condition value.
    #[serde(deserialize_with = "deserialize_weather_condition")]
    pub condition: WeatherCondition,
    /// Matches the WeatherKit symbol name value.
    pub symbol_name: String,
    /// Matches the WeatherKit precipitation value.
    #[serde(deserialize_with = "deserialize_precipitation")]
    pub precipitation: Precipitation,
    /// Chance of precipitation as a fraction from 0 to 1.
    pub precipitation_chance: f64,
    /// Precipitation amount in metres; multiply by 1000 for millimetres.
    pub precipitation_amount: f64,
    /// Cloud cover as a fraction from 0 to 1.
    pub cloud_cover: f64,
    /// Dew point in degrees Celsius.
    pub dew_point: f64,
    /// Relative humidity as a fraction from 0 to 1.
    pub humidity: f64,
    /// Matches the WeatherKit is daylight value.
    pub is_daylight: bool,
    /// Sea-level air pressure in hectopascals (millibars).
    pub pressure: f64,
    /// Matches the WeatherKit pressure trend value.
    #[serde(deserialize_with = "deserialize_pressure_trend")]
    pub pressure_trend: PressureTrend,
    /// Matches the WeatherKit uv index value.
    pub uv_index: UVIndex,
    /// Visibility in metres.
    pub visibility: f64,
    /// Matches the WeatherKit wind value.
    pub wind: Wind,
    /// Snowfall amount in metres; multiply by 1000 for millimetres.
    #[serde(default)]
    pub snowfall_amount: Option<f64>,
    /// Matches the WeatherKit cloud cover by altitude value.
    #[serde(default)]
    pub cloud_cover_by_altitude: Option<CloudCoverByAltitude>,
}

impl HourForecast {
    /// Wraps the WeatherKit pressure value and trend.
    pub fn pressure_reading(&self) -> Pressure {
        Pressure::new(self.pressure, self.pressure_trend.clone())
    }
}

/// Wraps the WeatherKit hourly forecast payload.
#[derive(Debug, Clone, PartialEq, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct HourlyForecast {
    /// Matches the WeatherKit forecast value.
    pub forecast: Vec<HourForecast>,
    /// Matches the WeatherKit metadata value.
    pub metadata: WeatherMetadata,
}

impl HourlyForecast {
    pub(crate) fn from_owned_ptr(ptr: *mut c_void) -> Result<Self, WeatherKitError> {
        parse_json_from_handle(
            ptr,
            ffi::hourly_forecast::wk_hourly_forecast_release,
            ffi::hourly_forecast::wk_hourly_forecast_copy_json,
            "hourly forecast",
        )
    }

    /// Returns the number of WeatherKit hour forecasts in this collection.
    pub fn len(&self) -> usize {
        self.forecast.len()
    }

    /// Returns whether this WeatherKit hour forecast collection is empty.
    pub fn is_empty(&self) -> bool {
        self.forecast.is_empty()
    }

    /// Iterates over the WeatherKit hour forecasts in this collection.
    pub fn iter(&self) -> std::slice::Iter<'_, HourForecast> {
        self.forecast.iter()
    }
}

impl<'a> IntoIterator for &'a HourlyForecast {
    type Item = &'a HourForecast;
    type IntoIter = std::slice::Iter<'a, HourForecast>;

    fn into_iter(self) -> Self::IntoIter {
        self.iter()
    }
}
